tt_metal/impl/buffers/buffer.cpp

```diff
@@ -198,89 +227,158 @@ BufferPageMapping generate_buffer_page_mapping(const Buffer &buffer) {
     return buffer_page_mapping;
 }
 
-Buffer::Buffer(const Buffer &other) :
-    device_(other.device_),
-    size_(other.size_),
-    page_size_(other.page_size_),
-    buffer_type_(other.buffer_type_),
-    buffer_layout_(other.buffer_layout_),
-    shard_parameters_(other.shard_parameters_),
-    bottom_up_(other.bottom_up_),
-    buffer_page_mapping_(other.buffer_page_mapping_),
-    allocate_(other.allocate_) {
-    if (this->allocate_) {
-        this->allocate();
+void Buffer::allocate() {
+    {
+        std::unique_lock lock(allocation_mutex_);
+        TT_FATAL(allocation_status_.load(std::memory_order::relaxed) == AllocationStatus::NOT_ALLOCATED, "Can't allocate buffer after it was already allocated");
+        allocation_status_.store(AllocationStatus::ALLOCATION_REQUESTED, std::memory_order::relaxed);
```

Comment:
Did setting `std::memory_order` to default cause perf overheads? any perf overheads in general pushing async alloc/dealloc down to metal?

Reply:
It might be, but in this case it will never be visible.

Reply:
Yes, it's a theoretical overhead, unsure if it would be visible. I think it's easier to just specify `std::memory_order::relaxed` rather than keep wondering if the performance loss is significant.